File: Project/Simulator/Organisms/Animals/Animal.java
package Simulator.Organisms.Animals;

import Simulator.Tile;
import Utils.DynamicDirections;
import Simulator.Organisms.Organism;
import Simulator.World;
import Simulator.GlobalSettings;

import org.json.JSONObject;

public abstract class Animal extends Organism {
    protected Tile oldTile = null;

    public void move(Tile newTile) {
        if (newTile == null)
            return;
        oldTile = this.tile;
        this.tile = newTile;
        this.tile.placeOrganism(this);
        oldTile.removeOrganism(this);
    }

    protected void move(DynamicDirections direction) {
        move(this.tile.getNeighbour(direction));
    }

    public Animal(int power, int initiative, World world, Type type) {
        super(type, power, initiative, world);
    }

    public Animal(JSONObject json, World world) {
        super(json, world);
        int index = json.getInt("old_tile_index");
        oldTile = index == -1 ? null : world.getTile(index);
    }

    @Override
    public void action() {
        move(this.tile.getRandomNeighbour());
    }

    @Override
    public void collision(Organism other) {
        if (other.collisionReaction(this))
            return;
        if (this.getClass() == other.getClass()) {
            this.undoMove();
            if (!GlobalSettings.AI_REPRODUCE)
                return;
            other.skipTurn();
            if (getBreedCooldown() > 0)
                return;
            Tile newTile = other.getTile().getRandomFreeNeighbour();
            if (newTile == null)
                return;
            Animal newAnimal = construct();
            newAnimal.skipTurn();
            this.setBreedCooldown(5);
            other.setBreedCooldown(5);
            newAnimal.setBreedCooldown(10);
            world.addOrganism(newAnimal, newTile);

            world.addLog(this.getSymbol() + " and " + other.getSymbol() +
                    " bred a new " + newAnimal.getSymbol() + "!");

        } else if (getPower() > other.getPower()) {
            other.die();
            world.addLog(this.getSymbol() + " killed " + other.getSymbol() + "!");
        } else {
            this.die();
            world.addLog(this.getSymbol() + " was killed by " +
                    other.getSymbol() + "!");
        }
    }

    @Override
    public boolean collisionReaction(Organism other) {
        return false;
    }

    public void setBreedCooldown(int cooldown) {
        this.reproductionCooldown = cooldown;
    }

    public int getBreedCooldown() {
        return this.reproductionCooldown;
    }

    public void undoMove() {
        if (oldTile == null)
            return;
        oldTile.placeOrganism(this);
        tile.removeOrganism(this);
        tile = oldTile;
        oldTile = null;
    }

    public abstract Animal construct();

    @Override
    public JSONObject toJson() {
        JSONObject json = super.toJson();
        int index = oldTile == null ? -1 : oldTile.index;
        json.put("old_tile_index", index);
        return json;
    }
}

File: Project/Simulator/Organisms/Animals/Animal.py
import json

from abc import ABC, abstractmethod
from tkinter import N

from Simulator.Tile import Tile
from Simulator.Organisms.Organism import Organism
from Simulator.GlobalSettings import GlobalSettings
from Utils.DynamicDirections import DynamicDirections


class Animal(Organism, ABC):
    def __init__(self, power, initiative, world, type, json=None):
        super().__init__(type, power, initiative, world, json)
        self._oldTile = None
        if json is not None:
            index = json["old_tile_index"]
            self._oldTile = world.getTile(index) if index != -1 else None

    def move(self, newTileOrDirection):
        if (newTileOrDirection == None):
            return
        if isinstance(newTileOrDirection, DynamicDirections):
            newTile = self._tile.getNeighbour(newTileOrDirection)
        else:
            newTile = newTileOrDirection
        if (newTile == None):
            return
        self._oldTile = self._tile
        self._tile = newTile
        self._tile.placeOrganism(self)
        if self._oldTile is not None:
            self._oldTile.removeOrganism(self)

    def undoMove(self):
        if self._oldTile is None:
            return
        self._oldTile.placeOrganism(self)
        self._tile.removeOrganism(self)
        self._tile = self._oldTile
        self._oldTile = None

    def action(self):
        self.move(self._tile.getRandomNeighbour())

    def getBreedCooldown(self):
        return self._reproductionCooldown

    def setBreedCooldown(self, value):
        self._reproductionCooldown = value

    def collision(self, other: Organism):
        if other.collisionReaction(self):
            return

        if type(self) == type(other):
            self.undoMove()
            if not GlobalSettings.AI_REPRODUCE:
                return
            other.skipTurn()
            if self.getBreedCooldown() > 0:
                return
            tile = other.getTile()
            new_Tile = tile.getRandomFreeNeighbour() if tile is not None else None
            if new_Tile is None:
                return
            new_animal = self.construct()
            if (new_animal is None):
                raise Exception(
                    "construct method must return an instance of Animal")
            new_animal.skipTurn()
            self.setBreedCooldown(5)
            other.setBreedCooldown(5)
            new_animal.setBreedCooldown(10)
            self._world.addOrganism(new_animal, new_Tile)

            self._world.addLog(
                f"{self.getSymbol()} and {other.getSymbol()} bred a new {new_animal.getSymbol()}!")
        elif self.getPower() > other.getPower():
            other.die()
            self._world.addLog(
                f"{self.getSymbol()} killed {other.getSymbol()}!")
        else:
            self.die()
            self._world.addLog(
                f"{self.getSymbol()} was killed by {other.getSymbol()}!")

    def collisionReaction(self, other):
        return False

    @abstractmethod
    def construct(self):
        pass

    def toJson(self):
        json = super().toJson()
        json["old_tile_index"] = self._oldTile.INDEX if self._oldTile is not None else -1
        return json
